fl4health/clients/dynamic_weight_exchange_client.py

```diff
@@ -0,0 +1,121 @@
+import copy
+from pathlib import Path
+from typing import Sequence
+
+import torch
+import torch.nn as nn
+from flwr.common.typing import Config, NDArrays
+
+from fl4health.clients.basic_client import BasicClient
+from fl4health.parameter_exchange.layer_exchanger import NormDriftParameterExchanger
+from fl4health.parameter_exchange.parameter_exchanger_base import ParameterExchanger
+from fl4health.utils.losses import LossMeterType
+from fl4health.utils.metrics import Metric, MetricMeterType
+
+
+class DynamicWeightExchangeClient(BasicClient):
+    def __init__(
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        metric_meter_type: MetricMeterType = MetricMeterType.AVERAGE,
+    ) -> None:
+        """
+        Dynamic weight exchange client used to exchange a dynamic subset of layers per client.
+
+        Args:
+            data_path (Path): path to the data to be used to load the data for client-side training
+            metrics (Sequence[Metric]): Metrics to be computed based on the labels and predictions of the client model
+            device (torch.device): Device indicator for where to send the model, batches, labels etc. Often 'cpu' or
+                'cuda'
+            loss_meter_type (LossMeterType, optional): Type of meter used to track and compute the losses over
+                each batch. Defaults to LossMeterType.AVERAGE.
+            metric_meter_type (MetricMeterType, optional): Type of meter used to track and compute the metrics over
+                each batch. Defaults to MetricMeterType.AVERAGE.
+            checkpointer (Optional[TorchCheckpointer], optional): Checkpointer to be used for client-side
+                checkpointing. Defaults to None.
+        """
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            metric_meter_type=metric_meter_type,
+        )
+        # Initial model parameters to be used in calculating weight shifts during training
+        self.initial_model: nn.Module
+        # Parameter exchanger to be used in server-client exchange of dynamic layers.
+        self.parameter_exchanger: NormDriftParameterExchanger
```

Comment:
One thought I had is that the dynamic weight exchanger could be more general than the norm drift parameter exchanger its constrained to using. We may want to define other parameter exchanger that select a certain set of weights each FL round to exchange according to some crtieria. However, lacking a concrete alternative parameter exchanger perhaps we can save it for a future work.

Reply:
Yes, we have a ticket in the backlog to tackle making this more generic! Just haven't gotten to it yet.